File: leetcode/easy/125-validPalindrome.java
// first pass solution
class Solution {
    public boolean isPalindrome(String s) {
        s = s.replaceAll("[^a-zA-Z0-9]", "").toLowerCase();
        String[] arr = s.split("");
        
        int j = arr.length - 1;
        
        for (int i = 0; i < arr.length / 2; i++) {
            String temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            j--;
        }
        return String.join("", arr).equals(s);
    }
}

// second pass solution
class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder sb = new StringBuilder();
        
        for (char c : s.toCharArray()) {
            if (Character.isDigit(c) || Character.isLetter(c)) {
                sb.append(Character.toLowerCase(c));
            }
        }
        return sb.toString().equals(sb.reverse().toString());
    }
}

File: leetcode/easy/125-validPalindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0;
        int j = s.length() - 1;

        while (i <= j) {
            while (i < s.length() && i <= j && !iswalnum(s[i])) i++;
            while (j >= 0 && j >= i && !iswalnum(s[j])) j--;
            if (i > j) return true;
            if (tolower(s[i]) != tolower(s[j])) return false;
            i++;
            j--;
        }

        return true;
    }
};
